docs/developer/best-practice.md

```diff
@@ -22,258 +41,86 @@ Finally, we have delivery commitments. Larger more intentional refactors may be
 
-> Interfaces in Go provide a way to specify the behavior of an object: if something can do this, then it can be used here. - [Effective Go](https://go.dev/doc/effective_go#interfaces_and_types)
+Package level state is rarely required. If it is, it should be composed of re-usable constructs defined and exposed from the package itself.
 
-Factory functions are not to be confused with initializing constructors. Factory functions typically return abstractions (an interface) satisfied by 1 of several implementations. A factory could be as simple as:
+See https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_package_level_state).
 
-```go
-// Foo and Bar satisfy fmt.Stringer
-func NewStringer(key string) (fmt.Stringer, error) {
-    switch key {
-    case "foo":
-        return Foo{}, nil
-    case "bar":
-        return Bar{}, nil
-    default:
-        return nil, errors.New("unknown key")
-    }
-}
-```
+### "Accept interfaces, return structs"
 
-Factories generally shouldn't contain complex construction logic. That's the job of the initializing constructor for the concrete type.
+This section is included because the phrase surfaces regularly.
 
-##### Loggers
+Code Review Comments provides a simple explanation that covers the general case. However, we want to stress this is the general case, not all cases.
 
-Loggers are sometimes treated as a special dependency by making entire projects depend on a package level logger. This _masks_ a crucial dependency that is hard to change without impacting the rest of the code base. To understand why this is bad, consider extracting a package into its own repository for reusability. Depending on a package level logger, or any other package level dependencies, makes the extraction process unnecessarily difficult.
+When developing abstractions it may become necessary to return interfaces but this should be considered orthogonal to accepting interfaces.
 
-Logger's are not special dependencies, inject them like any other dependency. If you want to minimize the surface area of the default initializing constructor (`New...` funcs) you may construct noop loggers to satisfy the logger dependency and provide a means to override the logger with a setter.
+See https://github.com/golang/go/wiki/CodeReviewComments#interfaces
 
-### Testing
-
-Software development is costly and hazardous. Tests provide example usage, guard rails, and assurances. Without tests, you cannot have high quality code. For this reason, all code we spend time designing and writing is worth testing. Similarly, if it can be unit tested, it should be unit tested. Unit tests are the first line of defense and can be complimented with additional integration or end-to-end tests.
+### Unit tests
 
 Unit tests should:
 
 - focus on the public API of a package.
-- use the `_test` idiom for testing files.
+- prefer the `_test` when testing the public API of the package.
```

Comment:
curious why lowering it from "use" to "prefer"

Reply:
Its to remove contradictory statements. We say 'use' but then we also say 'if somethings complex and unexported feel free to not use and do another thing that could include putting the tests in the package.

This was most if not all of my motivation.

Reply:
Another though, specifically calling out the public API test may mean 'use' is appropriate.